Firmware/MotorControl/crc.hpp

```diff
@@ -0,0 +1,66 @@
+#ifndef __CRC_HPP
+#define __CRC_HPP
+
+#include <limits.h>
+
+// Default CRC-8 Polynomial: x^8 + x^5 + x^4 + x^2 + x + 1
+// Can protect a 4 byte payload against toggling of up to 5 bits
+//  source: https://users.ece.cmu.edu/~koopman/crc/index.html
+constexpr uint8_t CRC8_DEFAULT = 0x37;
+
+// Default CRC-16 Polynomial: 0x9eb2 x^16 + x^13 + x^12 + x^11 + x^10 + x^8 + x^6 + x^5 + x^2 + 1
+// Can protect a 135 byte payload against toggling of up to 5 bits
+//  source: https://users.ece.cmu.edu/~koopman/crc/index.html
+// Also known as CRC-16-DNP
+constexpr uint16_t CRC16_DEFAULT = 0x3d65;
+
+// Calculates an arbitrary CRC for one byte.
+// Adapted from https://barrgroup.com/Embedded-Systems/How-To/CRC-Calculation-C-Code
+template<typename T, unsigned POLYNOMIAL>
+static T calc_crc(T remainder, uint8_t value) {
+    constexpr T BIT_WIDTH = (CHAR_BIT * sizeof(T));
+    constexpr T TOPBIT = (1 << (BIT_WIDTH - 1));
```

Comment:
Signed overflow if BIT_WIDTH is 32 or 64. Use `1ull << (BIT_WIDTH - 1)` or `(T)1 << (BIT_WIDTH - 1)` instead.

Reply:
Good catch, fixed. For T=int32_t I don't see any issue though.

Reply:
I think in the int32_t case, the signed overflow is technically undefined (though in practice it's bit-wise equivalent to unsigned overflow, which is defined)